### DISCUSSION TITLE
How to post dictionary data type in FastAPI Request Forms ?

### DISCUSSION THREAD

**Discussion by srcodesrepo** [fastapi/fastapi#6322] [Questions] [answered]
### First Check

- [X] I added a very descriptive title to this issue.
- [X] I used the GitHub search to find a similar issue and didn't find it.
- [X] I searched the FastAPI documentation, with the integrated search.
- [X] I already searched in Google "How to X in FastAPI" and didn't find any information.
- [X] I already read and followed all the tutorial in the docs and didn't find an answer.
- [X] I already checked if it is not related to FastAPI but to [Pydantic](https://github.com/samuelcolvin/pydantic).
- [X] I already checked if it is not related to FastAPI but to [Swagger UI](https://github.com/swagger-api/swagger-ui).
- [X] I already checked if it is not related to FastAPI but to [ReDoc](https://github.com/Redocly/redoc).

### Commit to Help

- [X] I commit to help with one of those options 👆

### Example Code

```python
@menu.post("/add-menu")
async def addmenu(
    ItemName: str = Form(...),
    ItemShortDesc: str = Form(...),
    ItemPrice: str = Form(...),
    ItemCategory: list = Form(...),
    ItemStatus: str = Form(...),
    ItemChoices: list[dict[str,str]] = Form(...)):
    
    return ItemChoices
```

### Description

**POST Method using FastAPI Request Form**
Request Body | application/x-www-form-urlencoded

I am trying to do,

1. "list" of "dict" using request form - error Code:422
2. Only "dict" using request form - error Code:422
3. without "dict" - Its Worked

I got the below error, error Code:422

"detail": [
    {
      "loc": [
        "body",
        "ItemChoices",
        0
      ],
      "msg": "value is not a valid dict",
      "type": "type_error.dict"
    }
  ]
}

How can I use "dict" or "list" of "dict" in FastAPI request form? Please help me out of any possible chance.

### Operating System

Linux

### FastAPI Version

0.73.0

### Python Version

python3.9

### Additional Context

<img width="1425" alt="Screenshot 2022-02-12 at 8 55 06 AM" src="https://user-images.githubusercontent.com/88054701/153696030-1a889630-2f51-4290-a0c4-4dfa5d04e2f6.png">

**Comment by rogerbrinkmann** [comment]
@growwithsundar , I would suggest using a pydantic model for validating the form:

```python
from typing import List, Dict
from fastapi import FastAPI
from fastapi.param_functions import Depends
from pydantic import BaseModel

app = FastAPI()

class CustomForm(BaseModel):
    ItemName: str
    ItemShortDesc: str
    ItemPrice: float
    ItemCategory: List[str]
    ItemStatus: str
    ItemChoices: List[Dict[str, str]]


@app.post("/add-menu")
async def addmenu(form_data: CustomForm = Depends()):
    return form_data
```

With the code above, I get the following:
![image](https://user-images.githubusercontent.com/5690226/153754358-092c9e62-ec91-4411-99e6-9e529406e1a6.png)


And:
![image](https://user-images.githubusercontent.com/5690226/153754389-59dd5384-74bf-4237-8cec-6bd4db766678.png)

**Comment by srcodesrepo** [comment]
Thanks for your time and workaround. 
@rogerbrinkmann I tried already using Pydantic and its worked for me too.

When I send request from IOS Swift 5 using application/json request body, FastAPI throws error and I deep searched for solution but I couldn't find one proper solutions. Please help me on this and actually there are many queries regarding this in web and no proper solution for the same. Please help me and people who facing the similar issues.

**Sending Data From IOS Swift 5 **

```
var json: [String: Any] = [
"ItemName":txtItemName.text!.trimmingCharacters(in: .whitespacesAndNewlines), "ItemShortDesc":txtItemName.text!.trimmingCharacters(in: .whitespacesAndNewlines), "ItemImage":strItemMenuUrl, "ItemPrice":txtPrice.text!.trimmingCharacters(in: .whitespacesAndNewlines), "ItemCategory":arrCatagory.joined(separator: ","),
"ItemStatus":strStatus,
"ItemChoices": ["title":"Choice", “subtitle":"SubChoice"]]
```

**Adding Header value**

```
var headers: HTTPHeaders = [
            "Content-Type": "application/json"
	    "Accept": "application/json"
        ]
```

**Alamofire API Call**

```
AF.request
(url,
method: .post,
parameters: json,
encoding: URLEncoding.httpBody,
headers: headers
).responseJSON { response in})
```

**Response Error 422 from FastAPI**

```
{
    "detail": [
        {
            "loc": [
                "body",
                0
            ],
            "msg": "Expecting value: line 1 column 1 (char 0)",
            "type": "value_error.jsondecode",
            "ctx": {
                "msg": "Expecting value",
                "doc": "ItemCategory=Salads%2CHot%20Dishes&ItemChoices%5B%5D=Choice%20Title&ItemImage=https%3A//fb110h.deta.dev/download?name%3D2e477271272244d33fa61644579019x42472.png&ItemName=Chicken%20Roast&ItemPrice=10&ItemShortDesc=Chicken%20Roast&ItemStatus=1",
                "pos": 0,
                "lineno": 1,
                "colno": 1
            }
        }
    ]
}
```

**Comment by rogerbrinkmann** [comment]
@growwithsundar, using pydantic is not a workaround, but the preferred way to do this. 
But your issue is not related to FastAPI. The Swift app is sending a field called ItemImage `"ItemImage":strItemMenuUrl`.
Your first example code, didn't contain that field. So, you need to modify my example so that it matches what Swift is sending.

**Comment by upeshjindal** [comment, marked as answer]
@growwithsundar You cannot send the data as a json in form fields. It needs to be sent as a **stringified JSON** and on the FastAPI side, you need convert this to your desired object.

Form fields work okay for simple datatypes.

See the link for reference: https://stackoverflow.com/questions/60127234/how-to-use-a-pydantic-model-with-form-data-in-fastapi

**Comment by srcodesrepo** [comment]
> @growwithsundar You cannot send the data as a json in form fields. It needs to be sent as a **stringified JSON** and on the FastAPI side, you need convert this to your desired object.
> 
> 
> 
> Form fields work okay for simple datatypes.
> 
> 
> 
> See the link for reference: https://stackoverflow.com/questions/60127234/how-to-use-a-pydantic-model-with-form-data-in-fastapi

Thanks it worked
